libs/x-content/src/main/java/org/elasticsearch/common/xcontent/ParsedMediaType.java

```diff
@@ -0,0 +1,139 @@
+ * @see MediaType
+ * @see MediaTypeRegistry
+ */
+public class ParsedMediaType {
+    // TODO this should be removed once strict parsing is implemented https://github.com/elastic/elasticsearch/issues/63080
+    // sun.net.www.protocol.http.HttpURLConnection sets a default Accept header if it was not provided on a request.
+    // For this value Parsing returns null.
+    public static final String DEFAULT_ACCEPT_STRING = "text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2";
+
+    private final String type;
+    private final String subType;
+    private final Map<String, String> parameters;
+    // tchar pattern as defined by RFC7230 section 3.2.6
+    private static final Pattern TCHAR_PATTERN = Pattern.compile("[a-zA-z0-9!#$%&'*+\\-.\\^_`|~]+");
+
+    private ParsedMediaType(String type, String subType, Map<String, String> parameters) {
+        this.type = type;
+        this.subType = subType;
+        this.parameters = Collections.unmodifiableMap(parameters);
+    }
+
+    /**
+     * The parsed mime type without the associated parameters. Will always return lowercase.
+     */
+    public String mediaTypeWithoutParameters() {
+        return type + "/" + subType;
+    }
+
+    public Map<String, String> getParameters() {
+        return parameters;
+    }
+
+    /**
+     * Parses a header value into it's parts.
+     * Note: parsing can return null, but it will throw exceptions once https://github.com/elastic/elasticsearch/issues/63080 is done
+     * Do not rely on nulls
+     *
+     * @return a {@link ParsedMediaType} if the header could be parsed.
+     * @throws IllegalArgumentException if the header is malformed
+     */
+    public static ParsedMediaType parseMediaType(String headerValue) {
+        if (DEFAULT_ACCEPT_STRING.equals(headerValue) || "*/*".equals(headerValue)) {
+            return null;
+        }
+        if (headerValue != null) {
+            final String[] elements = headerValue.toLowerCase(Locale.ROOT).split("[\\s\\t]*;");
+
+            final String[] splitMediaType = elements[0].split("/");
+            if ((splitMediaType.length == 2 && TCHAR_PATTERN.matcher(splitMediaType[0].trim()).matches()
+                && TCHAR_PATTERN.matcher(splitMediaType[1].trim()).matches()) == false) {
+                throw new IllegalArgumentException("invalid media type [" + headerValue + "]");
+            }
+            if (elements.length == 1) {
+                return new ParsedMediaType(splitMediaType[0].trim(), splitMediaType[1].trim(), Collections.emptyMap());
+            } else {
+                Map<String, String> parameters = new HashMap<>();
+                for (int i = 1; i < elements.length; i++) {
+                    String paramsAsString = elements[i].trim();
+                    if (paramsAsString.isEmpty()) {
+                        continue;
+                    }
+                    // intentionally allowing to have spaces around `=` (RFC disallows this)
```

Comment:
The only mention I saw in the spec was " No whitespace is allowed between the header field-name and colon. " 

The field value is not covered by this mandate:
```
header-field   = field-name ":" OWS field-value OWS
```
This does not explicitly state that the field-value can not have whitespace. For example, "application/\<space\>json" isn't forbidden by the HTTP spec (spaces are allowed in the field-value). The correctness of spaces in mimetypes or parameters would be defined by https://tools.ietf.org/html/rfc6838 (media type spec). While it is pretty clear that "application/\<space\>json" != "application/json", I think we should still allow it since that is a nightmare to troubleshoot. Also the spec is not clear on if ;\<space\>param\<space\>=\<space\>value is acceptable. I think it actually is. 

I would suggest to remove the "RFC disallows this".

Reply:
by RFC I meant https://tools.ietf.org/html/rfc7231#section-3.1.1.1 where it refers to the rfc6838 (the one you quoted) and makes it more strict  
```
 Internet media types ought to be registered with IANA according to
   the procedures defined in [BCP13].

      Note: Unlike some similar constructs in other header fields, media
      type parameters do not allow whitespace (even "bad" whitespace)
      around the "=" character.
```

```BCP13 ->    [BCP13]    Freed, N., Klensin, J., and T. Hansen, "Media Type
              Specifications and Registration Procedures", BCP 13,
              RFC 6838, January 2013.```

Reply:
I missed that, sorry. You were right, probably shouldn't be lenient here, but only with the whitespace around the `=`.  The code here is fine, but a follow up (or on this PR) to remove adhere to that would probably be best. sorry for steering you in the wrong direction on this.

Reply:
I will add validation around `=`